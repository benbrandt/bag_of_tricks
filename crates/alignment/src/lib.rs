#![deny(clippy::all)]
#![warn(clippy::pedantic)]

use std::{convert::TryFrom, f64::consts::E, fmt};

use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};

fn exp_weight(val: usize) -> f64 {
    E.powi(i32::try_from(val).unwrap_or_default())
}

#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, PartialEq, Serialize)]
pub enum Attitude {
    Chaotic,
    Lawful,
    Neutral,
}

impl Attitude {
    fn weight(self, influences: &[Self]) -> f64 {
        exp_weight(influences.iter().filter(|&i| i == &self).count())
    }
}

#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, PartialEq, Serialize)]
pub enum Morality {
    Evil,
    Good,
    Neutral,
}

impl Morality {
    fn weight(self, influences: &[Self]) -> f64 {
        exp_weight(influences.iter().filter(|&i| i == &self).count())
    }
}

/// Trait to describe how this entity affects a character's alignment
pub trait AlignmentInfluences {
    /// List of attitude influences
    fn attitude(&self) -> Vec<Attitude> {
        vec![]
    }

    /// List of morality influences
    fn morality(&self) -> Vec<Morality> {
        vec![]
    }
}

/// Character alignment, both attitude and morality
#[derive(Clone, Deserialize, Serialize)]
pub struct Alignment(pub Attitude, pub Morality);

impl Alignment {
    /// Generate alignment, weighted by influences from other choices on the character sheet
    ///
    /// # Panics
    ///
    /// Will panic if weighting logic is wrong or there are no attitudes/moralities at all
    pub fn gen(
        rng: &mut impl Rng,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
    ) -> Self {
        let attitude = *Attitude::iter()
            .collect::<Vec<_>>()
            .choose_weighted(rng, |a| a.weight(attitude_influences))
            .unwrap();
        let morality = *Morality::iter()
            .collect::<Vec<_>>()
            .choose_weighted(rng, |a| a.weight(morality_influences))
            .unwrap();
        Self(attitude, morality)
    }

    /// Weight of a particular alignment based on influences.
    /// Useful for comparing things like deities.
    #[must_use]
    pub fn weight(
        &self,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
    ) -> f64 {
        self.0.weight(attitude_influences) + self.1.weight(morality_influences)
    }
}

impl AlignmentInfluences for Alignment {
    fn attitude(&self) -> Vec<Attitude> {
        vec![self.0]
    }

    fn morality(&self) -> Vec<Morality> {
        vec![self.1]
    }
}

impl fmt::Display for Alignment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self(Attitude::Neutral, Morality::Neutral) => write!(f, "Neutral"),
            Self(a, m) => write!(f, "{} {}", a, m),
        }
    }
}
