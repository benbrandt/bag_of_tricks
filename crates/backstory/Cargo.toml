[package]
name = "backstory"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
