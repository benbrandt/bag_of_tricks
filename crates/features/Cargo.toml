[package]
name = "features"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2018"
resolver = "2"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
citation = { path = "../citation" }
serde = { version = "1", features = ["derive"] }

[dev-dependencies]
insta = "1"
serde_yaml = "0.8"
