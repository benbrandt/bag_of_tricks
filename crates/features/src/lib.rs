use std::fmt;

use citation::Citation;
use serde::{Deserialize, Serialize};

/// A feature or trait a character has.
#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct Feature {
    /// Name of the feature or trait.
    pub title: &'static str,
    /// Citation for where more information about this feature is available.
    pub citation: Citation,
}

impl fmt::Display for Feature {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} ({})", self.title, self.citation)
    }
}

/// Trait for objects that provide features to a character.
pub trait Features {
    // Return a list of features this thing provides
    fn features(&self) -> Vec<Feature> {
        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use citation::Book;

    #[test]
    fn test_display() {
        let feature = Feature {
            title: "Title",
            citation: Citation(Book::Phb, 1),
        };
        insta::assert_snapshot!(format!("{}", feature));
    }
}
