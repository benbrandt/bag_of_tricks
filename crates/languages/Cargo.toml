[package]
name = "languages"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
deities = { path = "../deities" }
rand = "0.8"
rand_pcg = "0.3"
serde = { version = "1", features = ["derive"] }
strum = { version = "0.22", features = ["derive"] }
