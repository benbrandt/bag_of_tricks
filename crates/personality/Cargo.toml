[package]
name = "personality"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
alignment = { path = "../alignment" }
serde = { version = "1", features = ["derive"] }
strum = { version = "0.25", features = ["derive"] }
