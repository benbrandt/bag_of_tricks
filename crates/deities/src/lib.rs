#![deny(clippy::all)]
#![warn(clippy::pedantic)]

mod dragon;
mod dragonlance;
mod dwarven;
mod eberron;
mod elven;
mod forgotten_realms;
mod giant;
mod gnomish;
mod goblinoid;
mod greyhawk;
mod halfling;
mod historical;
mod kobold;
mod lizardfolk;
mod orc;

use std::fmt;

use alignment::{Alignment, Attitude, Morality};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::Display;

use self::{
    dragon::Dragon,
    dragonlance::Dragonlance,
    dwarven::{Duergar, Dwarven},
    eberron::Eberron,
    elven::{Drow, Elven},
    forgotten_realms::ForgottenRealms,
    giant::Giant,
    gnomish::Gnomish,
    goblinoid::{Bugbear, Goblin},
    greyhawk::Greyhawk,
    halfling::Halfling,
    historical::{Celtic, Egyptian, Greek, Norse},
    kobold::Kobold,
    lizardfolk::Lizardfolk,
    orc::Orc,
};

#[derive(Copy, Clone, Deserialize, Display, PartialEq, Serialize)]
pub enum Domain {
    Arcana,
    Death,
    Forge,
    Grave,
    Knowledge,
    Life,
    Light,
    Nature,
    Tempest,
    Trickery,
    War,
}

#[derive(Clone, Deserialize, Serialize)]
pub struct Deity<'a> {
    pub name: &'a str,
    pub titles: Vec<&'a str>,
    pub alignment: Alignment,
    pub domains: Vec<Domain>,
    pub symbols: Vec<&'a str>,
}

impl Deity<'_> {
    fn weight(&self, attitude_influences: &[Attitude], morality_influences: &[Morality]) -> f64 {
        self.alignment
            .weight(attitude_influences, morality_influences)
    }
}

impl<'a> fmt::Display for Deity<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "CHOSEN DEITY: {}", self.name)?;
        writeln!(f, "Titles: {}", self.titles.join("; "))?;
        writeln!(f, "Alignment: {}", self.alignment)?;
        writeln!(
            f,
            "Domains: {}",
            self.domains
                .iter()
                .map(|d| format!("{}", d))
                .collect::<Vec<_>>()
                .join(", ")
        )?;
        writeln!(f, "Symbols: {}", self.symbols.join(", "))
    }
}

pub(crate) trait Deities<'a> {
    fn deities() -> Vec<Deity<'a>>;
}

#[derive(Copy, Clone, Deserialize, Display, PartialEq, Serialize)]
pub enum Pantheon {
    Bugbear,
    Celtic,
    Dragon,
    Dragonlance,
    Drow,
    Duergar,
    Dwarven,
    Eberron,
    Egyptian,
    Elven,
    #[strum(serialize = "Forgotten Realms")]
    ForgottenRealms,
    Giant,
    Gnomish,
    Goblin,
    Greek,
    Greyhawk,
    Halfling,
    Kobold,
    Lizardfolk,
    Norse,
    Orc,
    None,
}

impl Pantheon {
    fn all_deities<'a>(self) -> Vec<Deity<'a>> {
        match self {
            Self::Bugbear => Bugbear::deities(),
            Self::Celtic => Celtic::deities(),
            Self::Dragon => Dragon::deities(),
            Self::Dragonlance => Dragonlance::deities(),
            Self::Drow => Drow::deities(),
            Self::Dwarven => Dwarven::deities(),
            Self::Duergar => Duergar::deities(),
            Self::Eberron => Eberron::deities(),
            Self::Egyptian => Egyptian::deities(),
            Self::Elven => Elven::deities(),
            Self::ForgottenRealms => ForgottenRealms::deities(),
            Self::Giant => Giant::deities(),
            Self::Gnomish => Gnomish::deities(),
            Self::Goblin => Goblin::deities(),
            Self::Greek => Greek::deities(),
            Self::Greyhawk => Greyhawk::deities(),
            Self::Halfling => Halfling::deities(),
            Self::Kobold => Kobold::deities(),
            Self::Lizardfolk => Lizardfolk::deities(),
            Self::Norse => Norse::deities(),
            Self::Orc => Orc::deities(),
            Self::None => vec![],
        }
    }

    #[must_use]
    pub fn deities<'a>(self, domain: Option<Domain>) -> Vec<Deity<'a>> {
        domain.map_or_else(
            || self.all_deities(),
            |d| {
                self.all_deities()
                    .into_iter()
                    .filter(|deity| deity.domains.contains(&d))
                    .collect::<Vec<_>>()
            },
        )
    }

    /// # Panics
    ///
    /// Will panic if no Pantheons available
    pub fn choose(
        rng: &mut impl Rng,
        addl_pantheons: Vec<Self>,
        domain: Option<Domain>,
        required: bool,
    ) -> Self {
        let mut options = vec![
            Self::ForgottenRealms,
            Self::Celtic,
            Self::Dragonlance,
            Self::Eberron,
            Self::Egyptian,
            Self::Greek,
            Self::Greyhawk,
            Self::Norse,
        ];
        options.extend(addl_pantheons);
        if !required {
            options.push(Self::None);
        }
        options.dedup();
        *domain
            .map_or(options.clone(), |_| {
                options
                    .into_iter()
                    .filter(|p| p == &Self::None || !p.deities(domain).is_empty())
                    .collect::<Vec<_>>()
            })
            .choose_weighted(rng, |p| p.weight())
            .unwrap()
    }

    pub fn choose_deity<'a>(
        self,
        rng: &mut impl Rng,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
        domain: Option<Domain>,
    ) -> Option<Deity<'a>> {
        self.deities(domain)
            .choose_weighted(rng, |d| d.weight(attitude_influences, morality_influences))
            .ok()
            .cloned()
    }

    #[must_use]
    pub fn weight(self) -> f64 {
        match self {
            Self::Bugbear
            | Self::Dragon
            | Self::Drow
            | Self::Duergar
            | Self::Dwarven
            | Self::Elven
            | Self::ForgottenRealms
            | Self::Giant
            | Self::Gnomish
            | Self::Goblin
            | Self::Halfling
            | Self::Kobold
            | Self::Lizardfolk
            | Self::Orc
            | Self::None => 100.0,
            // Require some sort of plane shift for your character to end up in the forgotten realms
            Self::Celtic
            | Self::Dragonlance
            | Self::Eberron
            | Self::Egyptian
            | Self::Greek
            | Self::Greyhawk
            | Self::Norse => 1.0,
        }
    }
}

pub trait Pantheons {
    fn addl_pantheons(&self) -> Vec<Pantheon> {
        vec![]
    }

    fn deity_required(&self) -> bool {
        false
    }
}
