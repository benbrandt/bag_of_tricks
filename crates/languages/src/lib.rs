use deities::{Pantheon, PantheonWeight, Pantheons};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};

#[derive(Clone, Copy, Debug, Deserialize, Display, EnumIter, PartialEq, Serialize)]
pub enum LanguageType {
    Exotic,
    Standard,
}

impl LanguageType {
    fn weight(self) -> f64 {
        match self {
            Self::Exotic => 1.0,
            Self::Standard => 2.0,
        }
    }
}

/// Available languages for a character to learn.
#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub enum Language {
    Abyssal,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gith,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

impl Language {
    // Sometimes you end up with dupes. Consume and replace with a new option
    pub fn gen(
        rng: &mut impl Rng,
        current_languages: &[Language],
        (amount, language_type): (usize, Option<LanguageType>),
    ) -> Vec<Self> {
        language_type
            .map_or_else(
                || Self::iter().collect::<Vec<Self>>(),
                |lt| {
                    let filtered = Self::iter()
                        .filter(|l| l.language_type() == lt)
                        .collect::<Vec<Self>>();
                    if filtered.is_empty() {
                        Self::iter().collect::<Vec<Self>>()
                    } else {
                        filtered
                    }
                },
            )
            .iter()
            .filter(|l| !current_languages.contains(l))
            .copied()
            .collect::<Vec<_>>()
            .choose_multiple_weighted(rng, amount, |l| l.language_type().weight())
            .unwrap()
            .copied()
            .collect()
    }

    fn language_type(self) -> LanguageType {
        match self {
            Self::Common
            | Self::Dwarvish
            | Self::Elvish
            | Self::Giant
            | Self::Gnomish
            | Self::Goblin
            | Self::Halfling
            | Self::Orc => LanguageType::Standard,
            Self::Abyssal
            | Self::Celestial
            | Self::DeepSpeech
            | Self::Draconic
            | Self::Gith
            | Self::Infernal
            | Self::Primordial
            | Self::Sylvan
            | Self::Undercommon => LanguageType::Exotic,
        }
    }
}

impl Pantheons for Language {
    fn addl_pantheons(&self) -> Vec<(Pantheon, PantheonWeight)> {
        (match self {
            Language::Abyssal
            | Language::Celestial
            | Language::DeepSpeech
            | Language::Gith
            | Language::Infernal
            | Language::Primordial
            | Language::Sylvan => vec![],
            Language::Common => vec![Pantheon::ForgottenRealms],
            Language::Draconic => vec![Pantheon::Dragon, Pantheon::Kobold, Pantheon::Lizardfolk],
            Language::Dwarvish => vec![Pantheon::Dwarven],
            Language::Elvish => vec![Pantheon::Elven],
            Language::Giant => vec![Pantheon::Giant],
            Language::Gnomish => vec![Pantheon::Gnomish],
            Language::Goblin => vec![Pantheon::Bugbear, Pantheon::Goblin],
            Language::Halfling => vec![Pantheon::Halfling],
            Language::Orc => vec![Pantheon::Orc],
            Language::Undercommon => vec![Pantheon::Drow, Pantheon::Duergar],
        })
        .into_iter()
        .map(|p| (p, PantheonWeight::Possible))
        .collect::<Vec<_>>()
    }
}

/// Trait for describing what languages are provided by a given object.
pub trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Additional languages this character can learn.
    fn addl_languages(&self) -> (usize, Option<LanguageType>) {
        (0, None)
    }
}
