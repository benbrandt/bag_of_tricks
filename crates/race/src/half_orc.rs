use std::fmt;

use alignment::{AlignmentInfluences, Attitude, Morality};
use attack::Resistances;
use backstory::Backstory;
use characteristics::{
    in_inches, names::Name, AgeRange, Appearance, CharacteristicDetails, Characteristics,
    HeightAndWeightTable, Size, Speed, WeightMod,
};
use citation::{Book, Citation, CitationList, Citations};
use deities::{Pantheon, PantheonWeight, Pantheons};
use dice_roller::{Die, RollCmd};
use features::{Feature, Features};
use languages::{Language, Languages};
use personality::{Influence, PersonalityOptions};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use stats::{
    ability::{AbilityScore, AbilityScoreType, Skill},
    proficiencies::{Proficiencies, Proficiency},
};
use trinkets::Trinkets;

use super::{
    human::Human,
    orc::{Orc, BONDS, FLAWS, IDEALS, TRAITS},
    Race,
};

const HEIGHT_AND_WEIGHT: HeightAndWeightTable = HeightAndWeightTable {
    base_height: in_inches(4, 10),
    base_weight: 140,
    height_mod: RollCmd(2, Die::D10),
    weight_mod: WeightMod::Roll(RollCmd(2, Die::D6)),
};

#[derive(Default, Deserialize, Serialize)]
pub struct HalfOrc;

impl AlignmentInfluences for HalfOrc {
    fn attitude(&self) -> Vec<Attitude> {
        vec![Attitude::Chaotic]
    }

    fn morality(&self) -> Vec<Morality> {
        vec![Morality::Evil]
    }
}

impl Appearance for HalfOrc {}

impl Backstory for HalfOrc {}

impl Characteristics for HalfOrc {
    fn get_age_range(&self) -> AgeRange {
        AgeRange(8..=75)
    }

    fn get_base_speeds(&self) -> Vec<Speed> {
        vec![Speed::Walking(30)]
    }

    fn get_height_and_weight_table(&self) -> &HeightAndWeightTable {
        &HEIGHT_AND_WEIGHT
    }

    fn get_size(&self) -> Size {
        Size::Medium
    }

    fn has_human_ancestry(&self) -> bool {
        true
    }
}

impl Citations for HalfOrc {
    fn citations(&self) -> CitationList {
        CitationList(vec![Citation(Book::Phb, 40)])
    }
}

impl Features for HalfOrc {
    fn features(&self) -> Vec<Feature> {
        vec![
            // Thanks to your orc blood, you have superior vision in dark and dim conditions. You can see in dim light within 60 feet of you as if it were bright light, and in darkness as if it were dim light. You can't discern color in darkness, only shades of gray.
            Feature {
                title: "Darkvision",
                citation: Citation(Book::Phb, 41),
            },
            // When you are reduced to 0 hit points but not killed outright, you can drop to 1 hit point instead. You can't use this feature again until you finish a long rest.
            Feature {
                title: "Relentless Endurance",
                citation: Citation(Book::Phb, 41),
            },
            // When you score a critical hit with a melee weapon attack, you can roll one of the weapon's damage dice one additional time and add it to the extra damage of the critical hit.
            Feature {
                title: "Savage Attacks",
                citation: Citation(Book::Phb, 41),
            },
        ]
    }
}

impl Languages for HalfOrc {
    fn languages(&self) -> Vec<Language> {
        vec![Language::Common, Language::Orc]
    }
}

impl Name for HalfOrc {
    /// First name can be either orc or human name
    fn gen_name(&self, rng: &mut impl Rng, characteristics: &CharacteristicDetails) -> String {
        let first_name = *[
            Human::gen_first_name(rng, characteristics),
            Orc::gen_first_name(rng, characteristics),
        ]
        .choose(rng)
        .unwrap();
        let last_name = *[
            Human::gen_surname(rng, characteristics),
            Orc::gen_epithet(rng),
        ]
        .choose(rng)
        .unwrap();
        format!("{} {}", first_name, last_name)
    }
}

impl Pantheons for HalfOrc {
    fn addl_pantheons(&self) -> Vec<(Pantheon, PantheonWeight)> {
        vec![(Pantheon::Orc, PantheonWeight::Likely)]
    }
}

impl PersonalityOptions for HalfOrc {
    fn bonds(&self) -> Vec<String> {
        BONDS.iter().map(|&s| s.to_string()).collect()
    }

    fn flaws(&self) -> Vec<String> {
        FLAWS.iter().map(|&s| s.to_string()).collect()
    }

    fn ideals(&self) -> Vec<(String, Influence)> {
        IDEALS.iter().map(|&(s, i)| (s.to_string(), i)).collect()
    }

    fn traits(&self) -> Vec<String> {
        TRAITS.iter().map(|&s| s.to_string()).collect()
    }
}

impl Proficiencies for HalfOrc {
    fn proficiencies(&self) -> Vec<Proficiency> {
        vec![Proficiency::Skill(Skill::Intimidation)]
    }
}

impl Race for HalfOrc {
    fn gen(_: &mut impl Rng) -> Self {
        Self
    }

    fn abilities(&self) -> Vec<AbilityScore> {
        vec![
            AbilityScore(AbilityScoreType::Constitution, 1),
            AbilityScore(AbilityScoreType::Strength, 2),
        ]
    }
}

impl Resistances for HalfOrc {}

impl Trinkets for HalfOrc {}

impl fmt::Display for HalfOrc {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Half-Orc")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_snapshot() {
        let mut rng = Pcg64::seed_from_u64(1);
        let half_orc = HalfOrc::gen(&mut rng);
        insta::assert_yaml_snapshot!(half_orc);
    }

    #[test]
    fn test_snapshot_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let half_orc = HalfOrc::gen(&mut rng);
        insta::assert_snapshot!(format!("{}", half_orc));
    }

    #[test]
    fn test_snapshot_abilities() {
        let half_orc = HalfOrc;
        insta::assert_yaml_snapshot!(half_orc.abilities());
    }

    #[test]
    fn test_snapshot_citations() {
        let half_orc = HalfOrc;
        insta::assert_yaml_snapshot!(half_orc.citations());
    }

    #[test]
    fn test_snapshot_features() {
        let half_orc = HalfOrc;
        insta::assert_yaml_snapshot!(half_orc.features());
    }

    #[test]
    fn test_snapshot_addl_pantheons() {
        let half_orc = HalfOrc;
        insta::assert_yaml_snapshot!(half_orc.addl_pantheons());
    }

    #[test]
    fn test_bonds() {
        let half_orc = HalfOrc;
        insta::assert_yaml_snapshot!(half_orc.bonds());
    }

    #[test]
    fn test_flaws() {
        let half_orc = HalfOrc;
        insta::assert_yaml_snapshot!(half_orc.flaws());
    }

    #[test]
    fn test_ideals() {
        let half_orc = HalfOrc;
        insta::assert_yaml_snapshot!(half_orc.ideals());
    }

    #[test]
    fn test_traits() {
        let half_orc = HalfOrc;
        insta::assert_yaml_snapshot!(half_orc.traits());
    }
}
