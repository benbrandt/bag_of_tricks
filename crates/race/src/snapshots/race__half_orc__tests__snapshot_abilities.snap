---
source: crates/race/src/half_orc.rs
expression: half_orc.abilities()

---
- - Constitution
  - 1
- - Strength
  - 2

